kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,750 @@
+local cjson = require("cjson.safe")
+local utils = require("kong.resty.dns_client.utils")
+local mlcache = require("kong.resty.mlcache")
+local resolver = require("resty.dns.resolver")
+
+local now       = ngx.now
+local log       = ngx.log
+local ERR       = ngx.ERR
+local WARN      = ngx.WARN
+local ALERT     = ngx.ALERT
+local timer_at  = ngx.timer.at
+
+local type          = type
+local pairs         = pairs
+local ipairs        = ipairs
+local math_min      = math.min
+local math_random   = math.random
+local table_insert  = table.insert
+
+local parse_hosts   = utils.parse_hosts
+local ipv6_bracket  = utils.ipv6_bracket
+local search_names  = utils.search_names
+local get_round_robin_answers          = utils.get_round_robin_answers
+local get_weighted_round_robin_answers = utils.get_weighted_round_robin_answers
+
+local req_dyn_hook_run_hooks = require("kong.dynamic_hook").run_hooks
+local cycle_aware_deep_copy  = require("kong.tools.utils").cycle_aware_deep_copy
+
+
+-- Constants and default values
+
+local DEFAULT_ERROR_TTL   = 1     -- unit: second
+local DEFAULT_STALE_TTL   = 4
+local DEFAULT_EMPTY_TTL   = 30
+-- long-lasting TTL of 10 years for hosts or static IP addresses in cache settings
+local LONG_LASTING_TTL    = 10 * 365 * 24 * 60 * 60
+local STALE_UPDATE_DELAY  = 5
+
+local DEFAULT_ORDER = { "LAST", "SRV", "A", "AAAA", "CNAME" }
+
+local TYPE_SRV      = resolver.TYPE_SRV
+local TYPE_A        = resolver.TYPE_A
+local TYPE_AAAA     = resolver.TYPE_AAAA
+local TYPE_CNAME    = resolver.TYPE_CNAME
+local TYPE_LAST     = -1
+
+local valid_type_names = {
+  SRV     = TYPE_SRV,
+  A       = TYPE_A,
+  AAAA    = TYPE_AAAA,
+  CNAME   = TYPE_CNAME,
+  LAST    = TYPE_LAST,
+}
+
+local typstrs = {
+  [TYPE_SRV]      = "SRV",
+  [TYPE_A]        = "A",
+  [TYPE_AAAA]     = "AAAA",
+  [TYPE_CNAME]    = "CNAME",
+}
+
+local HIT_L3 = 3 -- L1 lru, L2 shm, L3 callback, L4 stale
+
+local hitstrs = {
+  [1] = "hit_lru",
+  [2] = "hit_shm",
+  [3] = "hit_cb",
+  [4] = "hit_stale",
+}
+
+-- server replied error from the DNS protocol
+local NAME_ERROR_CODE    = 3 -- response code 3 as "Name Error" or "NXDOMAIN"
+-- client specific error
+local CACHE_ONLY_EC      = 100
+local CACHE_ONLY_ESTR    = "cache only lookup failed"
+local CACHE_ONLY_ANSWERS = { errcode = CACHE_ONLY_EC, errstr = CACHE_ONLY_ESTR }
```

Comment:
We can protect this table with https://stevedonovan.github.io/Penlight/api/libraries/pl.tablex.html#readonly to avoid accidental modification.